provider/pkg/resources/cfn_custom_resource_test.go

```diff
@@ -0,0 +1,1540 @@
+				"serviceToken":       resource.NewStringProperty(serviceToken),
+				"bucket":             resource.NewStringProperty(bucketName),
+				"resourceType":       resource.NewStringProperty(resourceType),
+			}
+
+			err = c.Delete(ctx, urn, physicalResourceID, inputs, state, tt.timeout)
+
+			if tt.expectedError != "" {
+				require.Error(t, err)
+				assert.Contains(t, err.Error(), tt.expectedError)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestCfnCustomResource_Delete_PresignPutObjectFail(t *testing.T) {
+	t.Parallel()
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockLambdaClient := client.NewMockLambdaClient(ctrl)
+	mockS3Client := client.NewMockS3Client(ctrl)
+
+	expectedError := fmt.Errorf("failed to presign put object")
+	stackID := "stack-id"
+	serviceToken := "arn:aws:lambda:us-west-2:123456789012:function:my-function"
+	bucketKeyPrefix := "bucket-key-prefix"
+	bucketName := "bucket-name"
+	resourceType := "Custom::MyResource"
+	physicalResourceID := "physical-resource-id"
+
+	mockS3Client.EXPECT().PresignPutObject(gomock.Any(), bucketName, matchesBucketKeyPrefix(bucketKeyPrefix), gomock.Any()).Return("", expectedError)
+
+	c := &cfnCustomResource{
+		providerName: "testProvider",
+		lambdaClient: mockLambdaClient,
+		s3Client:     mockS3Client,
+	}
+	ctx := context.Background()
+	urn := urn.URN("urn:pulumi:testProject::test::aws-native:cloudformation:CfnCustomResource::dummy")
+
+	inputs := resource.PropertyMap{
+		"serviceToken":    resource.NewStringProperty(serviceToken),
+		"resourceType":    resource.NewStringProperty(resourceType),
+		"stackID":         resource.NewStringProperty(stackID),
+		"bucketName":      resource.NewStringProperty(bucketName),
+		"bucketKeyPrefix": resource.NewStringProperty(bucketKeyPrefix),
+		"customResourceProperties": resource.NewObjectProperty(resource.PropertyMap{
+			"key": resource.NewStringProperty("value"),
+		}),
+	}
+
+	state := resource.PropertyMap{
+		"physicalResourceId": resource.NewStringProperty(physicalResourceID),
+		"data":               resource.NewObjectProperty(resource.NewPropertyMapFromMap(map[string]interface{}{})),
+		"stackId":            resource.NewStringProperty(stackID),
+		"serviceToken":       resource.NewStringProperty(serviceToken),
+		"bucket":             resource.NewStringProperty(bucketName),
+		"resourceType":       resource.NewStringProperty(resourceType),
+	}
+
+	err := c.Delete(ctx, urn, physicalResourceID, inputs, state, 0)
+
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), expectedError.Error())
+}
+
+func TestCfnCustomResource_Delete_LambdaInvokeFail(t *testing.T) {
+	t.Parallel()
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockLambdaClient := client.NewMockLambdaClient(ctrl)
+	mockS3Client := client.NewMockS3Client(ctrl)
+
+	expectedError := fmt.Errorf("failed to invoke lambda")
+	stackID := "stack-id"
+	serviceToken := "arn:aws:lambda:us-west-2:123456789012:function:my-function"
```

Comment:
Here's the serviceToken example, I see. Indeed it looks like a lambda ARN. Interesting.